File: public/lib/fidl/examples/echo_client_cpp/echo_client.cc
// Copyright 2017 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "lib/app/cpp/application_context.h"
#include "lib/app/cpp/connect.h"
#include "lib/fidl/cpp/bindings/binding_set.h"
#include "lib/fsl/tasks/message_loop.h"
#include "lib/fxl/macros.h"

#include "lib/fidl/examples/services/echo.fidl.h"

namespace echo {

class ResponsePrinter {
 public:
  void Run(const fidl::String& value) const {
    printf("***** Response: %s\n", value.get().c_str());

    fsl::MessageLoop::GetCurrent()->QuitNow();
  }
};

class EchoClientApp {
 public:
  EchoClientApp()
    : context_(app::ApplicationContext::CreateFromStartupInfo()) {
    FXL_DCHECK(context_);
  }

  bool Start(std::string server_url, std::string msg) {
    auto launch_info = app::ApplicationLaunchInfo::New();
    launch_info->url = server_url;
    launch_info->services = echo_provider_.NewRequest();

    context_->launcher()->CreateApplication(std::move(launch_info),
                                            controller_.NewRequest());

    app::ConnectToService(echo_provider_.get(), echo_.NewRequest());
    FXL_DCHECK(echo_);

    echo_->EchoString(msg,
                      [this](fidl::String value) {
                        ResponsePrinter printer;
                        printer.Run(std::move(value));
                      });
    return true;
  }

 private:
  std::unique_ptr<app::ApplicationContext> context_;
  app::ServiceProviderPtr echo_provider_;
  app::ApplicationControllerPtr controller_;
  echo::EchoPtr echo_;
};

}  // namespace echo

int main(int argc, const char** argv) {
  std::string server_url = "file:///system/apps/echo_server_cpp";
  std::string msg = "hello world";

  for (int i = 1; i < argc - 1; ++i) {
    if (!strcmp("-u", argv[i])) {
      server_url = argv[++i];
    } else if (!strcmp("-m", argv[i])) {
      msg = argv[++i];
    }
  }
  fsl::MessageLoop loop;

  echo::EchoClientApp app;
if (app.Start(server_url, msg))
    loop.Run();
  return 0;
}

File: examples/fidl/echo_client_go/echo_client.go
// Copyright 2018 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package main

import (
	"flag"
	"fmt"

	"app/context"
	"fidl/bindings"
	"svc/services"

	"garnet/public/lib/app/fidl/application_controller"
	"garnet/public/lib/app/fidl/application_launcher"

	"garnet/examples/fidl/services/echo"
)

type echoClientApp struct {
	ctx          *context.Context
	echoProvider *services.Provider
	controller   *application_controller.ApplicationController_Pointer
	echo         *echo.Echo_Proxy
}

func (a *echoClientApp) start(serverURL string, msg string) {
	pr, err := a.echoProvider.NewRequest()
	if err != nil {
		fmt.Println(err)
		return
	}

	launchInfo := application_launcher.ApplicationLaunchInfo{
		Url:              serverURL,
		DirectoryRequest: pr}

	cr, cp := a.controller.NewRequest()
	a.ctx.Launcher.CreateApplication(launchInfo, cr)
	a.controller = cp
	defer a.controller.Close()

	r, p := a.echo.NewRequest(bindings.GetAsyncWaiter())
	a.echoProvider.ConnectToService(r)
	a.echo = p
	defer a.echo.Close()

	response, err := a.echo.EchoString(&msg)
	if err != nil {
		fmt.Println(err)
	} else {
		fmt.Println("client:", *response)
	}
}

func main() {
	serverURL := flag.String("server", "echo_server_go", "server URL")
	msg := flag.String("m", "Hello, Go World", "message")

	flag.Parse()

	a := &echoClientApp{
		ctx:          context.CreateFromStartupInfo(),
		echoProvider: services.NewProvider()}

	a.start(*serverURL, *msg)
}
